### ISSUE TITLE
Native Integration: Menu API

### ISSUE THREAD

**Issue by bryphe** [revery-ui/revery#250]
Like Electron's [`Menu` API](https://electronjs.org/docs/api/menu), it'd be very useful from Revery to create native application menus.

The Menu API from electron is a good starting point, but I wonder if we could use `<JSX />` syntax to create a more React-like experience for creating these menu items. In particular, one thing I found awkward about the Electron API was updating menu items. I believe we could streamline this experience and simplify the API a bit.

**Comment by wokalski** [comment]
I think this should be a common library shared by brisk and revery. cc @rauanmayemir

**Comment by rauanmayemir** [comment]
Indeed. You could actually depend on `brisk-macos` and have `Menu` available. Here's an example how to use them https://github.com/briskml/brisk/blob/master/tester-macos/bin/Menu.re.

They don't support updates, but that could easily be improved.

Previously, I've had them as a lib like `brisk-cocoa-menu`, but then decided to move bingings in because it sucked to maintain them separately.

**Comment by bryphe** [comment]
Makes sense to share the native components out and have a consistent API surface. I was initially considering a `revery-native` library to host some of these. 

Some 'native' APIs I'd like to expose from it:
- `Menu` (tracked by this issue - and sounds like already a lot of work done on the MacOS side!)
- Accessibility (TBD)
- 'Tray' integration
- Desktop notifications
- System dialogs
- Potentially text input / IME

**Comment by wokalski** [comment]
@rauanmayemir note that revery has support for all platforms

**Comment by bryphe** [comment]
BTW it'd be fine to host this in another place too - if you wanted to host this in brisk, maybe a `brisk-platform` for these common elements (ie, factoring out the `Menu` API, etc) would be a good starting point. Didn't mean to imply it all has to live in Revery 😄 Just a common place for us to share some of these bindings.

It's great if the API for these pieces is transferable - I could imagine someone prototyping in Revery, and then finalizing their platform-specific apps in Brisk.

**Comment by rauanmayemir** [comment]
> It's great if the API for these pieces is transferable between brisk/revery - I could imagine someone prototyping in Revery, and then finalizing their platform-specific apps in Brisk.

I think this should drive our priorities. 👍

On the one hand, Brisk is aiming to target platform-specific native apis and provide bindings for it and it would make sense for Revery to use it under the hood as an implementation detail. But on the other hand, we ourselves as Brisk would probably prefer higher-level api to be the same for end-users.

I'm positively open to having a common api package. It could go either way, but we'd still need to split up native bindings per-platform as separate sub-packages (unless we use dune variants, but they're not available yet).

I'm open to suggestions.

**Comment by wokalski** [comment]
I don't care where is what. What I'd like is a correct structure - Cross platform Menu.mli would be in module `CrossPlatform` (let's say) but it would be built upon lower level platform specific primitives from `System` modules which would compile to different native bindings behind the scenes using CPPO. So that we have the highest common denominator of the various platform specific (system) APIs in the cross platform package. So we would have the same interface but native menus.

**Comment by rauanmayemir** [comment]
I'm not sure CPPO is the way to go. 😕

**Comment by wokalski** [comment]
In other words we’re gonna have:
- Menu.mli

And then in the link time it’s gonna be filled with Menu_macos.Compat on the Mac Menu_windows.Compat on windows etc.
